src/main/webapp/placeGuides/Location.js

```diff
@@ -3,40 +3,54 @@
  * coordinates or by a place_id, which will be decoded with PlacesService.
  */
 class Location {
-  constructor(position, mapsPlace) {
-    this._mapsPlace = mapsPlace;
-    if (this._mapsPlace != null) {
-      this._position = this._mapsPlace.geometry.location;
-    } else {
-      this._position = position;
-    }
+  constructor(latitude, longitude, placeId) {
+    this._position = new google.maps.LatLng(latitude, longitude);
+    this._placeId = placeId;
+    this._placeName = undefined;
   }
 
   get position() {
     return this._position;
   }
 
-  get mapsPlace() {
-    return this._mapsPlace;
+  get placeId() {
+    return this._placeId;
   }
 
-  static constructLocationBasedOnCoordinates(positionLat, positionLng) {
-    return new Location(
-        new google.maps.LatLng(positionLat, positionLng), null);
+  get placeName() {
+    if (this._placeName !== undefined) {
+      const thisPlaceName = this._placeName;
+      return new Promise(function(resolve, reject) {
+        resolve(thisPlaceName);
+      });
+    } else {
+      if (this._placeId === undefined) {
+        return new Promise(function(resolve, reject) {
+          resolve(undefined);
+        });
+      }
+      const thisLocation = this;
+      return this.definePlaceName()
+          .then((newPlaceName) => {
+            thisLocation._placeName = newPlaceName;
+            return newPlaceName;
+          });
+    }
```

Comment:
If feels a little strange that a getter returns a Promise. In my mind getters usually return just what they have locally. Or is this common in Javascript?

Reply:
I converted it to a general function(not a getter).